kinoml/core/protein.py

```diff
@@ -0,0 +1,350 @@
+            path = Path(path)
+            try:
+                # TODO: Workaround the getstate/setstate calls?
+                loaded = pmd.load_file(path)
+                state = loaded.__getstate__()
+                protein = cls()
+                protein.__setstate__(state)
+                return protein
+            except Exception as e:
+                ext = path.suffix
+        return cls._from_dispatcher(path, ext, "ext", "_from_file_", **kwargs)
+
+    @classmethod
+    def from_identifier(cls, identifier, resource, **kwargs):
+        return cls._from_dispatcher(
+            identifier, resource, "resource", "_from_id_", **kwargs
+        )
+
+    @classmethod
+    def from_sequence(cls, sequence, **kwargs):
+        # TODO: Add atoms to residues
+        protein = cls()
+        for symbol in sequence:
+            r = pmd.residue.AminoAcidResidue.get(symbol)
+            protein.residues.add(r)
+        return protein
+
+    @classmethod
+    def _from_id_uniprot(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_hgnc(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_name(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @property
+    def sequence(self):
+        s = "".join([r.symbol for r in self.residues])
+        return AminoAcidSequence(s)
+
+
+class Kinase(Protein):
+
+    """
+    Extends ``Protein`` to provide kinase-specific methods of
+    instantiation.
+    """
+
+    @classmethod
+    def _from_id_klifs(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_kinmap(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+    @classmethod
+    def _from_id_manning(cls, identifier, **kwargs):
+        raise NotImplementedError
+
+
+class Biosequence(str):
+    """
+    Base class for string representations of biological polymers
+    (nucleic acids, peptides, proteins...)
+    TODO: How to handle several mutations at the same time, while
+          keeping indices relevant (after a deletion, a replacement
+          or insertion position might be wrong).
+    """
+
+    ALPHABET = set(ascii_letters)
+    _ACCESSION_URL = None
+    ACCESSION_MAX_RETRIEVAL = 50
+
+    def __new__(cls, value, header="", _provenance=None, *args, **kwargs):
+        diff = set(value).difference(cls.ALPHABET)
+        if diff:
+            raise ValueError(
+                f"Biosequence can only contain characters in {cls.ALPHABET}, "
+                f"but found these extra ones: {diff}."
+            )
+        s = super().__new__(cls, value, *args, **kwargs)
+        s.header = header
+        s._provenance = {}
+        # TODO: We might override some provenance data with this blind update
+        if _provenance is not None:
+            s._provenance.update(_provenance)
+        return s
+
+    @classmethod
+    def from_accession(cls, *accessions):
```

Comment:
If this convenience constructor is only for NCBI identifiers, let's call it `from_ncbi()`.